ecommerce/extensions/payment/views/fomopay.py

```diff
@@ -0,0 +1,376 @@
+
+class FomopayQRView(View):
+    """Starts FOMO Pay payment process.
+
+    This view acts as an interface between the user and FOMO Pay,
+    it'll mainly display the QR code and instructions on how to make a payment with the WeChat app.
+    """
+
+    # Disable CSRF validation. The internal POST requests to render this view
+    # don't include the CSRF token as hosted-side payment processor are
+    # excepted to be externally hosted, but this is not the case.
+    # Instead of changing the checkout flow for all the payment processors
+    # this view is marked as exempt.
+    @method_decorator(csrf_exempt)
+    def dispatch(self, request, *args, **kwargs):
+        return super(FomopayQRView, self).dispatch(request, *args, **kwargs)
+
+    def post(self, request):
+        """Render QR Code.
+
+        Process the incoming request to render the QR in the template.
+        """
+        try:
+            qr_url = self._get_qr_link(request)
+            qr_img = self._generate_qr(qr_url)
+            order_id = request.POST.get('transaction')
+            status_url = reverse('fomopay:status')
+
+            context = {
+                'qrcode': qr_img,
+                'order_id': order_id,
+                'status_url': status_url,
+                'receipt_page': self._get_receipt_page(order_id),
+                'error_page': reverse('payment_error'),
+            }
+
+            return render(request, 'payment/fomopay.html', context)
+        except:  # pylint: disable=bare-except
+            logger.error('FOMO Pay QR Code generation failed for request[%s].',
+                         request.POST.dict())
+            return redirect(reverse('payment_error'))
+
+    def _generate_qr(self, qr_url):
+        """
+        Create a base64 representation of the QR code image.
+        """
+        img = qrcode.make(qr_url)
+        img_io = StringIO.StringIO()
+        img.save(img_io, format='PNG')
+
+        return base64.b64encode(img_io.getvalue())
+
+    def _get_qr_link(self, request):
+        """
+        Process the outgoing request to the FOMO API to obtain the
+        link needed to create the QR code
+        """
+        params = request.POST.dict()
+        api_url = params.pop('api_url')
+        response = requests.post(url=api_url, data=params)
+        fomo_response = response.json()
+        qr_url = fomo_response.get('url')
+
+        if not qr_url:
+            raise PaymentError
+
+        return qr_url
+
+    def _get_receipt_page(self, order_id):
+        """ Get receipt page for given order."""
+        receipt_page_url = get_receipt_page_url(
+            self.request.site.siteconfiguration,
+            order_number=order_id
+        )
+
+        return receipt_page_url
+
+
+class FomopayPaymentStatusView(APIView):
+    """Polls payment status.
+
+    Return the current status of the payment by polling the order
+    or the basket status.
+    """
+    # TODO: implement security to prevent unauthorized access.
```

Comment:
@felipemontoya I think is necessary to implement security here to prevent that other users poll orders that doesn't belong to them and to prevent DDoS attacks. What do you think would be the best way to do this? I was thinking using JWT tokens but then we need to provide another endpoint to get the token.

Reply:
answered through different channel